api/libproject/devstart.rb

```diff
@@ -1100,66 +1089,63 @@ def with_cloud_proxy_and_db_env(cmd_name, args)
-    raise("expected exactly one of CIRCLE_BRANCH and CIRCLE_TAG env vars to be set")
-  end
-  is_master = ENV.fetch("CIRCLE_BRANCH", "") == "master"
-  if !is_master and !ENV.has_key?("CIRCLE_TAG")
-    common.status "not master or a git tag, nothing to deploy"
-    return
-  end
-
-  unless Workbench::in_docker?
-    exec *(%W{docker run --rm -v #{File.expand_path("..")}:/w -w /w/api
-      allofustest/workbench:buildimage-0.0.9
-      ./project.rb #{cmd_name}} + args)
-  end
+def deploy(cmd_name, args)
+  ensure_docker cmd_name, args
 
-  common = Common.new
-  unless File.exist? "circle-sa-key.json"
-    common.error "Missing service account credentials file circle-sa-key.json."
-    exit 1
-  end
+  op = WbOptionsParser.new(cmd_name, args)
+  op.add_option(
+    "--account [account]",
+    lambda {|opts, v| opts.account = v},
+    "Service account to act as for deployment, if any. Defaults to the GAE " +
+    "default service account."
+  )
+  op.add_option(
+    "--version [version]",
+    lambda {|opts, v| opts.version = v},
+    "Version to deploy (e.g. your-username-test)"
+  )
+  op.add_validator lambda {|opts| raise ArgumentError unless opts.version}
+  op.add_option(
+    "--promote",
+    lambda {|opts, v| opts.promote = true},
+    "Promote this version to immediately begin serving API traffic"
+  )
+  op.add_option(
+    "--no-promote",
+    lambda {|opts, v| opts.promote = false},
+    "Deploy, but do not yet serve traffic from this version - DB migrations are still applied"
+  )
+  op.add_validator lambda {|opts| raise ArgumentError if opts.promote.nil?}
 
-  if is_master
-    common.status "Running database migrations..."
-    with_cloud_proxy_and_db_env(cmd_name, args) do |ctx|
-      migrate_database
-      load_config(ctx.project)
+  gcc = GcloudContextV2.new(op)
+  op.parse.validate
+  gcc.validate
 
-      common.status "Pushing GCS artifacts..."
-      deploy_gcs_artifacts(cmd_name, %W{--project #{ctx.project}})
-    end
-  end
+  common = Common.new
+  common.status "Running database migrations..."
+  with_cloud_proxy_and_db(gcc, service_account=op.opts.account) do |ctx|
+    migrate_database
+    load_config(ctx.project)
 
-  promote = ""
-  version = ""
-  if is_master
-    # Note that --promote will generally be a no-op, as we expect
-    # circle-ci-test to always be serving 100% traffic. Pushing to an existing
-    # live version will immediately make those changes live. In the event that
-    # someone mistakenly pushes a different version manually, this --promote
-    # will restore us to the expected circle-ci-test version on the next commit.
-    promote = "--promote"
-    version = "circle-ci-test"
-  elsif ENV.has_key?("CIRCLE_TAG")
-    promote = "--no-promote"
-    version = ENV["CIRCLE_TAG"]
+    common.status "Pushing GCS artifacts..."
+    deploy_gcs_artifacts(cmd_name, %W{--project #{ctx.project}})
+
+    # Keep the cloud proxy context open for the service account credentials.
```

Comment:
So are we using the service account credentials to deploy API and public-api? It looks like we're not for ui anymore?

Seems like either way might be fine, but we may want to be consistent?

Reply:
Done - see above